modules/lang-painless/src/main/java/org/elasticsearch/painless/Whitelist.java

```diff
@@ -0,0 +1,802 @@
+     *   an arrow symbol, the Java class name, a opening bracket, a newline, constructor/method/field
+     *   specifications, a closing bracket, and a final newline. Within a complex type the following
+     *   may be parsed:
+     *   -- A constructor may be specified starting with an opening parenthesis, followed by a
+     *      comma-delimited list of Painless type names corresponding to the type/class names for
+     *      the equivalent Java parameter types (these must be white-listed as well), a closing
+     *      parenthesis, and a newline.
+     *   -- A method may be specified starting with a Painless type name for the return type,
+     *      followed by the Java name of the method (which will also be the Painless name for the
+     *      method), an opening parenthesis, a comma-delimited list of Painless type names
+     *      corresponding to the type/class names for the equivalent Java parameter types
+     *      (these must be white-listed as well), a closing parenthesis, and a newline.
+     *   -- An augmented method may be specified starting with a Painless type name for the return
+     *      type, followed by the fully qualified Java name of the class the augmented method is
+     *      part of (this class does not need to be white-listed), the Java name of the method
+     *      (which will also be the Painless name for the method), an opening parenthesis, a
+     *      comma-delimited list of Painless type names corresponding to the type/class names
+     *      for the equivalent Java parameter types (these must be white-listed as well), a closing
+     *      parenthesis, and a newline.
+     *   -- A field may be specified starting with a Painless type name for the equivalent Java type
+     *      of the field, followed by the Java name of the field (which all be the Painless name
+     *      for the field), and a newline.
+     *
+     * Note there must be a one-to-one correspondence of Painless type names to Java type/class names.
+     * If the same Painless type is defined across multiple files and the Java class is the same, all
+     * specified constructors, methods, and fields will be merged into a single Painless type.  Dynamic
+     * types will always be equivalent to an {@link Object} as a base, but resolution of method calls
+     * and field access will be determined at run-time.
+     *
+     * The following example is used to create a single white-list text file:
+     *
+     * {@code
+     * # dynamic types
+     *
+     * dynamic def
+     *
+     * # primitive types
+     *
+     * class int -> int {
+     * }
+     *
+     * # complex types
+     *
+     * class Example -> my.package.Example {
+     *   # constructors
+     *   ()
+     *   (int)
+     *   (Example)
+     *
+     *   # method
+     *   Example add(int)
+     *   int add(Example, Example)
+     *   void example()
+     *
+     *   # augmented
+     *   Example some.other.Class sub(int)
```

Comment:
Can we use `.` here as the separator between class name and method? Then it looks just like a static method call (which it is).